src/viewport/services/auth_rate_limiter.py

```
@@ -0,0 +1,429 @@
+        raise InvalidClientAddressError("Forwarded headers disagree")
+    return forwarded or xff or ()
+
+
+def resolve_client_ip(request: Request, trusted_proxy_cidrs: str | Iterable[str | IPNetwork] = ()) -> str:
+    """Resolve a client IP without trusting headers from direct clients.
+
+    For a trusted immediate peer, the first address left of the trusted proxy
+    suffix is used. Every forwarded hop must be a valid IP address.
+    """
+    if request.client is None:
+        raise InvalidClientAddressError("Request has no socket peer")
```

Ruff (0.16.1)

[warning] Avoid specifying long messages outside the exception class

(TRY003)